File: apps/api/app/schemas/vdd.py
"""M22 — Value Delivery Document schemas.

Four sections (renewal-conversation source of truth):
  1. client_strategic_priorities  list[str]              — pillars / themes
  2. agreed_success_metrics       list[MetricSnapshot]   — usually mirrored from success_metrics
  3. beroes_approach              list[ApproachItem]     — per-initiative + 3-lever savings
  4. value_delivered              list[ValueDeliveredItem] — CSM-attributed $-amounts

`extra="allow"` so future fields (cadence overrides, escalation triggers,
exec-summary blurb) flow through without schema churn — same pattern as
M15 cs_goals and M19 success_contract.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Cost / risk / adoption — the 3-lever savings framework.
SavingsLever = Literal["cost", "risk", "adoption"]
# Mirrors cs_goals.initiatives stage vocabulary.
InitiativeStage = Literal[
    "proposed", "committed", "in_flight", "implemented", "blocked", "cancelled"
]


class MetricSnapshot(BaseModel):
    """Snapshot ref of one success_metric row (id + display fields)."""

    model_config = ConfigDict(extra="allow")

    id: str | None = None  # success_metrics.id (when sourced from M20)
    name: str
    target: str | None = None
    current: str | None = None
    status: Literal["green", "amber", "red", "grey"] | None = None


class ApproachItem(BaseModel):
    """Beroe's approach against one initiative + 3-lever savings."""

    model_config = ConfigDict(extra="allow")

    initiative_id: str | None = None  # cs_goals.initiatives[].id reference
    initiative_name: str
    approach: str | None = None
    levers: list[SavingsLever] = Field(default_factory=list)
    stage: InitiativeStage | None = None


class ValueDeliveredItem(BaseModel):
    """CSM-attributed $-amount rollup for one initiative."""

    model_config = ConfigDict(extra="allow")

    initiative_id: str | None = None
    initiative_name: str
    identified_musd: float | None = None
    committed_musd: float | None = None
    implemented_musd: float | None = None
    note: str | None = None


class VddOut(BaseModel):
    """What the API returns."""

    model_config = ConfigDict(extra="allow")

    account_id: UUID

    client_strategic_priorities: list[str] = Field(default_factory=list)
    agreed_success_metrics: list[MetricSnapshot] = Field(default_factory=list)
    beroes_approach: list[ApproachItem] = Field(default_factory=list)
    value_delivered: list[ValueDeliveredItem] = Field(default_factory=list)

    exec_summary: str | None = None

    locked_at: datetime | None = None
    locked_by: UUID | None = None

    auto_drafted: bool = False
    is_editable: bool = False


class VddUpdate(BaseModel):
    """Partial whole-document update. Refusing to PATCH a locked VDD —
    the route returns 409. Caller must unlock first."""

    model_config = ConfigDict(extra="allow")

    client_strategic_priorities: list[str] | None = None
    agreed_success_metrics: list[MetricSnapshot] | None = None
    beroes_approach: list[ApproachItem] | None = None
    value_delivered: list[ValueDeliveredItem] | None = None
    exec_summary: str | None = Field(None, max_length=4000)


class VddLock(BaseModel):
    """No payload required — server validates required minimums are met."""

    model_config = ConfigDict(extra="allow")

File: apps/web/src/types/vdd.ts
// Mirrors apps/api/app/schemas/vdd.py.

export type SavingsLever = "cost" | "risk" | "adoption";
export type InitiativeStage =
  | "proposed"
  | "committed"
  | "in_flight"
  | "implemented"
  | "blocked"
  | "cancelled";

export interface MetricSnapshot {
  id?: string | null;
  name: string;
  target?: string | null;
  current?: string | null;
  status?: "green" | "amber" | "red" | "grey" | null;
}

export interface ApproachItem {
  initiative_id?: string | null;
  initiative_name: string;
  approach?: string | null;
  levers: SavingsLever[];
  stage?: InitiativeStage | null;
}

export interface ValueDeliveredItem {
  initiative_id?: string | null;
  initiative_name: string;
  identified_musd?: number | null;
  committed_musd?: number | null;
  implemented_musd?: number | null;
  note?: string | null;
}

export interface Vdd {
  account_id: string;
  client_strategic_priorities: string[];
  agreed_success_metrics: MetricSnapshot[];
  beroes_approach: ApproachItem[];
  value_delivered: ValueDeliveredItem[];
  exec_summary: string | null;
  locked_at: string | null;
  locked_by: string | null;
  auto_drafted: boolean;
  is_editable: boolean;
}

export interface VddUpdate {
  client_strategic_priorities?: string[] | null;
  agreed_success_metrics?: MetricSnapshot[] | null;
  beroes_approach?: ApproachItem[] | null;
  value_delivered?: ValueDeliveredItem[] | null;
  exec_summary?: string | null;
}

export const LEVERS: SavingsLever[] = ["cost", "risk", "adoption"];
export const LEVER_LABELS: Record<SavingsLever, string> = {
  cost: "Cost",
  risk: "Risk",
  adoption: "Adoption",
};
export const LEVER_TONES: Record<SavingsLever, string> = {
  cost: "bg-beroe-green/15 text-beroe-green border-beroe-green/30",
  risk: "bg-beroe-amber/15 text-beroe-amber border-beroe-amber/40",
  adoption: "bg-beroe-blue/10 text-beroe-blue border-beroe-blue/30",
};

export const STAGES: InitiativeStage[] = [
  "proposed",
  "committed",
  "in_flight",
  "implemented",
  "blocked",
  "cancelled",
];

/** Rollup totals across value_delivered for the attribution bar. */
export function attributionTotals(rows: ValueDeliveredItem[]) {
  const sum = (k: keyof ValueDeliveredItem) =>
    rows.reduce((acc, r) => acc + (Number(r[k]) || 0), 0);
  return {
    identified: sum("identified_musd"),
    committed: sum("committed_musd"),
    implemented: sum("implemented_musd"),
  };
}

export function fmtMusd(n: number | null | undefined): string {
  if (n === null || n === undefined || Number.isNaN(n)) return "—";
  if (n === 0) return "$0";
  return `$${n.toFixed(2)}M`;
}
